ruby_programming/automated_testing/rspec_part_one_basics.md

```diff
@@ -0,0 +1,822 @@
+
+~~~ruby
+# spec/todo_list_spec.rb
+
+require "spec_helper"
+require "todo_list"
+
+RSpec.describe TodoList do
+
+	describe "name" do
+
+  end
+end
+~~~
+
+The describe method takes an argument that can either be a class name or a string. Typically you will pass the name of the class you are testing to the outer most example group and use strings for the nested example groups.
+
+We will get into why you'd want to use a class name for some example groups and strings for others later on. For now, the most important thing to note is that the argument is used to provide a label that describes the behaviour we will test within the example group.
+
+In our case the the top level example group describes all the tests within it will be for the `TodoList` class.
+
+The nested example group describes all tests within it will be for the `name` method on the `TodoList` class.
+
+Structuring your test files and providing descriptive labels like this makes it easy for you to read the test file and determine what is being tested within it.
+
+The describe method also takes a block ( `do..end`). As we have already seen, nested example groups can go within the block. But we can also put individual examples within the block. Or in other words, our actual tests.
+
+#### It blocks
+
+`it` blocks are used to define the individual examples within our example groups. Much like the describe method,  `it` is also a method that accepts an argument that allows us to express what we expect to happen in that individual test, for example `it "returns the users name"`.
+
+However unlike the `describe` method, the argument **must** always be a string and we cannot nest anything within an `it` block. This is because `it` blocks are reserved for our actual test code.
+
+Let's create a few `it` blocks which we can use to test the `name` method of the todo list:
+
+~~~ruby
+# spec/todo_list_spec.rb
+
+require "spec_helper"
+require "todo_list"
+
+RSpec.describe TodoList do
+
+  describe "name" do
+    it "returns the name with 0 tasks remaining" do
+    end
+
+    it "returns the name with 1 task remaining" do
+    end
+
+    it "returns the name with 10 tasks remaining" do
+    end
+  end
+end
+~~~
+
+To see all the structuring we have been doing in action we need to execute our tests. To execute the tests, open the terminal and run the `$ rspec` command from the root directory of the `odin_rspec` project directory.
+
+The output, as shown in the code snippet below uses the arguments we passed to our example groups and `it` blocks to output the structure of our tests to the terminal.
+
+~~~bash
+TodoList
+  name
+    returns the name with 1 task remaining
+    returns the name with 0 tasks remaining
+    returns the name with 10 tasks remaining
+
+Finished in 0.00247 seconds (files took 0.1574 seconds to load)
+3 examples, 0 failures
+~~~
```

Comment:
In following along with this lesson, my output is slightly different.

```bash
TodoList
  name
    returns the name with 1 task remaining
    returns the name with 0 tasks remaining
    returns the name with 10 tasks remaining

Top 3 slowest examples (0.00044 seconds, 6.6% of total time):
  TodoList name returns the name with 0 tasks remaining
    0.00023 seconds ./spec/todo_list_spec.rb:6
  TodoList name returns the name with 1 task remaining
    0.00011 seconds ./spec/todo_list_spec.rb:9
  TodoList name returns the name with 10 tasks remaining
    0.00009 seconds ./spec/todo_list_spec.rb:12

Finished in 0.00659 seconds (files took 0.14693 seconds to load)
3 examples, 0 failures
```

Having the extra lines with the test times is not a big deal, but I wonder if people will find this confusing since their output will be slightly different than the lesson's output. 

I have never removed the `=` from the `=begin` and `=end` in the spec_helper file, so I have never seen the times in my output. I went back in and commented out `config.profile_examples = 10` because I did not want to see the times in my output. What are the benefits to removing those `=`?

Reply:
This is a bit of an hangover from experimenting, the main goal was to get the tests to output in the documentation format instead of the dots. But I can just add a `--format documentation` flag to the .rspec file to do that instead and make this a lot simpler.

Cheers for bringing this up, I hated having to make them go into the spec_helper and mess around with those equals symbols 😄